polkadot/node/collation-generation/src/lib.rs

```diff
@@ -310,7 +310,39 @@ async fn process_primary_block<Context: SubsystemContext>(
 
 	let bundles = request_extract_bundles(block_hash, extrinsics, ctx.sender()).await.await??;
 
-	let execution_receipt = match (config.processor)(block_hash, bundles).await {
+	let header = {
+		let (tx, rx) = oneshot::channel();
+		ctx.send_message(ChainApiMessage::BlockHeader(block_hash, tx)).await;
```

Comment:
Why don't we request randomness here directly? It looks like the only reason we request header is to later make another request and get randomness out of it. Would make more sense to get randomness for block height instead.

Or will header be used for something else here?

Reply:
You can't because you can't access a specific header given the block number in the runtime.

Reply:
If you don't want to pull in more subspace deps, you have to do it via a runtime api instead of extracting it directly inside the subsystem.

I'm not satisfied the code quality, but I'd rather not spending more time on the optimization at the moment.